README.md
# neumorphic_container

Fully customisable `Neumorphic Containers` for your flutter projects.

<img src="https://raw.githubusercontent.com/AditiRavi/Neumorphic-Container/main/img2.jpeg" height="300em" /> <img src="https://raw.githubusercontent.com/AditiRavi/Neumorphic-Container/main/img4.jpeg" height="300em" /> <img src="https://raw.githubusercontent.com/AditiRavi/Neumorphic-Container/main/img6.jpeg" height="300em" />

## Getting Started
In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  neumorphic_container: "^0.0.1+2"
```

In your library add the following import:

```dart
import 'package:neumorphic_container/neumorphic_container.dart';
```

For help getting started with Flutter, view the online [documentation](https://flutter.io/).

## Usage
Simple Container:
For best results, set the background
color of a surrounding widget to match
the color you set for the container.

```
NeumorphicContainer(
              height: 150,
              width: 150,
              borderRadius: 10,
              primaryColor: Color(0xfff0f0f0),
              curvature: Curvature.flat,
            ),
```
<img src="https://raw.githubusercontent.com/AditiRavi/Neumorphic-Container/main/img3.jpeg" height="300em" />

Container with Border:

```
 NeumorphicContainer(
              height: 130,
              width: 130,
              borderRadius: 150,
              primaryColor: Color(0xff292929),
              //add border color and thickness
              borderColor: Colors.orange,
              borderThickness: 1,
              curvature: Curvature.concave,
              child: Icon(Icons.phone,color: Colors.orange,size: 30,),
            ),
```
<img src="https://raw.githubusercontent.com/AditiRavi/Neumorphic-Container/main/img5.jpeg" height="300em" />

Curvature

```dart
import 'package:flutter/material.dart';
import 'package:neumorphic_container/neumorphic_container.dart';

class Example extends StatefulWidget {
  @override
  _ExampleState createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  @override
  Widget build(BuildContext context) {
    Color color = Color(0xfff0f0f0);
    return Scaffold(
      backgroundColor: color,
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Padding(
              padding: const EdgeInsets.all(20.0),
              child: NeumorphicContainer(
                height: 150,
                width: 150,
                borderRadius: 150,
                depth: 20,
                primaryColor: color,
                //flat neumorphism design
                curvature: Curvature.flat,
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(20.0),
              child: NeumorphicContainer(
                height: 150,
                width: 150,
                borderRadius: 150,
                primaryColor: color,
                spread: 5,
                //convex neumorphism design
                curvature: Curvature.convex,
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(20.0),
              child: NeumorphicContainer(
                height: 150,
                width: 150,
                borderRadius: 150,
                primaryColor: color,
                //concave neumorphism design
                curvature: Curvature.concave,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```
<img src="https://raw.githubusercontent.com/AditiRavi/Neumorphic-Container/main/img4.jpeg" height="400em" />

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.

### Show some :heart: and star the repo to support the project


CHANGELOG.md
## [0.0.1+2] - June 27,2020.

* Documentation Changes.

## [0.0.1+1] - June 27,2020.

* Fixed formatting issues.

## [0.0.1] - June 26,2020.

* Initial release.





example/lib/main.dart
import 'package:example/example.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Demo Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key,  required this.title}) : super(key: key);


  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Example( ),
    );
  }
}


lib/src/neomorphism.dart
import 'package:flutter/material.dart';

enum Curvature { concave, convex, flat }

class NeumorphicContainer extends StatelessWidget {
  final double? height;
  final double? width;
  final int? depth;
  final Color? primaryColor;
  final Color? borderColor;
  final double? borderThickness;
  final double? spread;
  final Widget? child;
  final double? borderRadius;
  final BorderRadius? customRadius;
  final Curvature? curvature;

  NeumorphicContainer({
    this.child,
    this.height,
    this.width,
    this.depth,
    this.borderThickness,
    this.primaryColor,
    this.borderColor,
    this.spread,
    this.borderRadius,
    this.customRadius,
    this.curvature,
  });
 
  List<Color> smoothGradient(baseColor, depth) {
    return [
      baseColor,
      baseColor,
    ];
  }

  List<Color> concaveGradients(baseColor, depth) {
    return [
      _baseColor(baseColor, 0 - depth),
      _baseColor(baseColor, depth),
    ];
  }

  List<Color> convexGradients(baseColor, depth) {
    return [
      _baseColor(baseColor, depth),
      _baseColor(baseColor, 0 - depth),
    ];
  }

  Color _baseColor(Color color, amount) {
    Map colors = {"red": color.red, "green": color.green, "blue": color.blue};
    colors = colors.map((key, value) {
      if (value + amount < 0) return MapEntry(key, 0);
      if (value + amount > 255) return MapEntry(key, 255);
      return MapEntry(key, value + amount);
    });
    return Color.fromRGBO(colors["red"], colors["green"], colors["blue"], 1);
  }

  @override
  Widget build(BuildContext context) {
    final double? heightValue = height == null ? null : height;
    final double? widthValue = width == null ? null : width;
    final int? depthValue = depth == null ? 20 : depth;
    Color colorValue = primaryColor == null ? Color(0xFFf0f0f0) : primaryColor!;
    final double spreadValue = spread == null ? 6 : spread!;
    final bool eValue = false;
    BorderRadius? borderRadiusValue = borderRadius == null
        ? BorderRadius.zero
        : BorderRadius.circular(borderRadius!);
    final double? thicknessValue = borderThickness == null ? 0 : borderThickness;
    Border borderValue = borderColor == null
        ? Border.all(color: Colors.transparent, width: thicknessValue!)
        : Border.all(color: borderColor!, width: thicknessValue!);

    if (customRadius != null) {
      borderRadiusValue = customRadius;
    }
    final Curvature? curveTypeValue =
        curvature == null ? Curvature.flat : curvature;

    List<BoxShadow> shadowList = [
      BoxShadow(
          color: _baseColor(colorValue, eValue ? 0 - depthValue! : depthValue),
          offset: Offset(0 - spreadValue, 0 - spreadValue),
          blurRadius: spreadValue),
      BoxShadow(
          color: _baseColor(colorValue, eValue ? depthValue : 0 - depthValue!),
          offset: Offset(spreadValue, spreadValue),
          blurRadius: spreadValue)
    ];

    if (eValue) shadowList = shadowList.reversed.toList();
    if (eValue) colorValue = _baseColor(colorValue, 0 - depthValue ~/ 2);

    late List<Color> gradientColors;
    switch (curveTypeValue) {
      case Curvature.concave:
        gradientColors = concaveGradients(colorValue, depthValue);
        break;
      case Curvature.convex:
        gradientColors = convexGradients(colorValue, depthValue);
        break;
      case Curvature.flat:
        gradientColors = smoothGradient(colorValue, depthValue);
        break;
    }

    return Container(
      height: heightValue,
      width: widthValue,
      child: child,
      decoration: BoxDecoration(
          border: borderValue,
          borderRadius: borderRadiusValue,
          color: colorValue,
          gradient: LinearGradient(
              begin: Alignment.topLeft,
              end: Alignment.bottomRight,
              colors: gradientColors),
          boxShadow: shadowList),
    );
  }
}


example/lib/example.dart
import 'package:flutter/material.dart';
import 'package:neumorphic_container/neumorphic_container.dart';

class Example extends StatefulWidget {
  @override
  _ExampleState createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  @override
  Widget build(BuildContext context) {
    Color color = Color(0xfff0f0f0);
    return Scaffold(
      backgroundColor: color,
      body: SingleChildScrollView(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Padding(
              padding: const EdgeInsets.all(20.0),
              child: NeumorphicContainer(
                height: 150,
                width: 150,
                borderRadius: 150,
                depth: 20,
                primaryColor: color,
                //flat neumorphism design
                curvature: Curvature.flat,
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(20.0),
              child: NeumorphicContainer(
                height: 150,
                width: 150,
                borderRadius: 55,
                primaryColor: Color(0xff292929),
                borderColor: Colors.orange,
                borderThickness: 1,
                spread: 5,
                //convex neumorphism design
                curvature: Curvature.convex,
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(20.0),
              child: NeumorphicContainer(
                height: 250,
                width: 250,
                borderRadius: 75,
                primaryColor: color,
                //concave neumorphism design
                curvature: Curvature.concave,

                child: Padding(
                    padding: const EdgeInsets.all(60.0),
                    child: NeumorphicContainer(
                      height: 50,
                      width: 50,
                      borderRadius: 5,
                      primaryColor: color,
                      //concave neumorphism design
                      curvature: Curvature.convex,
                    )),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
